### DISCUSSION TITLE
Disabling variable name-mangling

### DISCUSSION THREAD

**Discussion by gogo9th** [babel/babel#13065] [General] [answered]
Hi,

I use the following transformation option:

```
      var options = {
        retainLines: true,
        compact: false,
        sourceType: 'unambiguous',
        sourceMaps: true,
        presets: ['@babel/preset-env'],
        plugins: [
          ["transform-es2015-modules-simple-commonjs",
            {
              'noMangle': true,
              'allowTopLevelThis': true,
              'strictMode': false
            },
          ]
        ]
      }
```

And I transform the following sample code: 

```
function a() { }
function b() {
   a = function() {
      a = 1
   };
}
```

Then I get a mangled variable name as follows:
```
function _a() {}
function b() {
   _a = function a() {
      _a = 1;
   };
}  
```

Q1. Most importantly, I need to disable this mangling, because I am transforming Javascript code in each <script> tag by using Babel, but if any script is mangled as the above and changes the variable name `"a"` to `"_a"`, then other script tags referring to the variable `"a"` crashes with an error, because "a" is undefined. How can I disable this mangling? If I need to modify the module's source code, which line should I modify? 

Q2. If name-mangling is really necessary, I need at least the top-level variable/function name to stay the same, while lower-level variable/function names can be mangled. How could I do this?

**Comment by gogo9th** [comment, marked as answer]
I disabled the above function name warping by commenting out two lines of code in `node_modules/@babel/plugin-transform-fu
nction-name/lib/index.js` as follows:

```
var _default = (0, _helperPluginUtils.declare)(api => {
  api.assertVersion(7);
  return {
    name: "transform-function-name",
    visitor: {
      FunctionExpression: {
        exit(path) {
          if (path.key !== "value" && !path.parentPath.isObjectProperty()) {
            //const replacement = (0, _helperFunctionName.default)(path);
            //if (replacement) path.replaceWith(replacement);
          }
        }

      },
```

Would this cause any error or inconsistency? 
And is there any more elegant way of accomplishing this disabling with configuring plug-ins?

**Reply by nicolo-ribaudo** [reply]
You can pass the `exclude` option to `@babel/preset-env` to completely disable that plugin.

**Reply by gogo9th** [reply]
Do you know the exact syntax for this? I tried it in various ways, but all failed...

**Reply by fedeci** [reply]
```json
{
  "presets": [
    [
      "@babel/preset-env",
      { "exclude": ["@babel/plugin-transform-function-name"] }
    ]
  ]
}
```
This should work.

**Reply by gogo9th** [reply]
This indeed works. Thanks very much!
